### DISCUSSION TITLE
[QUESTION] kvssink keeps dropping frames from an rtsp source

### DISCUSSION THREAD

**Discussion by floatingpurr** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#1008] [Q&A] [answered]
I'm trying to stream to KVS several rtsp flows (video only). 

I built my environment exactly how it is described here: https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#build. However, kvssink keeps dropping frames with this message: 

> DEBUG - Dropping frame with flag: 1536

and the KVS stream remains empty.

## Case 1: foo (not working)

Here is my command for the camera _foo_:

```sh
$ gst-launch-1.0 -v rtspsrc location="rtsp://user:pass@foo" short-header=TRUE ! rtph264depay ! h264parse ! kvssink stream-name="YourStreamName" storage-size=128 access-key="aaa" secret-key="aaa" aws-region="eu-central-1"
```
and here is the [output](https://gist.github.com/floatingpurr/306d7fefc0a3e8e5034b012af12253d8#file-foo-log)

Here is the `fakesink`:

```bash
$ gst-launch-1.0 rtspsrc location="rtsp://user:pass@foo" ! rtph264depay ! capsfilter caps="video/x-h264" ! fakesink
```
here is the [output](https://gist.github.com/floatingpurr/306d7fefc0a3e8e5034b012af12253d8#file-foo_fakesink-log)

Even in VLC, the video is properly streamed.

## Case 2: bar (working!)

**I tried the same command with another camera (_bar_) and in that case, all the things work:**

```sh
$ gst-launch-1.0 -v rtspsrc location="rtsp://user:pass@bar" short-header=TRUE ! rtph264depay ! h264parse ! kvssink stream-name="YourStreamName" storage-size=128 access-key="aaa" secret-key="aaa" aws-region="eu-central-1"
```
and here is the [output](https://gist.github.com/floatingpurr/306d7fefc0a3e8e5034b012af12253d8#file-bar-log)

## Case 3: baz (not working, again)

**Strangely enough, if I try with the camera baz (same model of camera bar), things do not work again (dropped frames).**

```sh
$ gst-launch-1.0 -v rtspsrc location="rtsp://user:pass@baz" short-header=TRUE ! rtph264depay ! h264parse ! kvssink stream-name="YourStreamName" storage-size=128 access-key="aaa" secret-key="aaa" aws-region="eu-central-1"
```

here is the [output](https://gist.github.com/floatingpurr/306d7fefc0a3e8e5034b012af12253d8#file-baz-log)

Here is the `fakesink` for this case
```sh
$ gst-launch-1.0 -v rtspsrc location="rtsp://user:pass@baz" short-header=TRUE ! rtph264depay ! h264parse ! fakesink
```

and here is the [output](https://gist.github.com/floatingpurr/306d7fefc0a3e8e5034b012af12253d8#file-baz_fakesink-log)

Again, baz streams properly to VLC.

Unfortunately, I can't access to any camera control panel. I can just consume the rtsp streams as they are.

I'd kindly ask you some help on how to troubleshoot this kind of problem.

**Comment by hassanctech** [comment]
The frame dropping message should be benign.  Here are the frame flags:

https://gstreamer.freedesktop.org/documentation/gstreamer/gstbuffer.html?gi-language=c#GstBufferFlags

So `1536 = 1024 + 512` meaning both `GST_BUFFER_FLAG_HEADER` and `GST_BUFFER_FLAG_MARKER` are set, and `1600 = 1024 + 512 + 64` meaning `GST_BUFFER_FLAG_HEADER, GST_BUFFER_FLAG_MARKER, GST_BUFFER_FLAG_DISCONT` are set.  These will be dropped but should not impact the stream, in fact I see this happens in the `bar` case which seems to be working fine.

I did notice this in your logs:

`Could not receive any UDP packets for 5.0000 seconds, maybe your firewall is blocking it. Retrying using a tcp connection.`

If you know you will not support UDP in order to avoid this latency you can set this property:  https://gstreamer.freedesktop.org/documentation/rtsp/rtspsrc.html?gi-language=c#rtspsrc:protocols to only `tcp` because if you leave it to the default it will *always* try UDP first which will incur latency for you.  

In theory fixing the above to TCP only with your network configuration shouldn't fix the issue but I wouldn't be surprised if there were some extra issues that show up when you're renegotiating the rtsp stream the camera so I would definitely suggest to make this change if you know you will not support UDP.

This likely will not fix your issue but I see your terminate the pipeline in the foo case after 13s.  Can you please try to run it for 1-2 minutes and see if we can get some error codes returned from the SDK to be printed in the logs?

**Comment by floatingpurr** [comment]
Thank you @hassanctech for your suggestions. 

Here is the longer log for the camera foo:

[foo.log](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/files/11370931/foo.log)

And here is the same log with the property `protocols=tcp`

[foo_tcp.log](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/files/11371011/foo_tcp.log)


Plese note that in both cases I get in Kinesis :

> No media found
> Verify that you are streaming to eu-central-1 and the fragment timecodes are correct, or download the Kinesis Video Streams Producer SDK to set up your device.

**Reply by jdelapla** [reply]
The logs show that you're dropping with flag 1536 very often. Can you check the dts and pts of the frames?
https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/63537aaa4398d4879c3f5a41d1149878cbf0dad1/src/gstreamer/gstkvssink.cpp#L1290

Add the pts and dts to this log line, we'll make the change in an upcoming release as well.

**Comment by floatingpurr** [comment]
I build the project from e99e7e94d897f309c773cb3bb677cab5c9b5f587. This is the corresponding section of the file you mentioned at that commit:

https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/e99e7e94d897f309c773cb3bb677cab5c9b5f587/src/gstreamer/gstkvssink.cpp#L1101-L1104

Do you mean something like this?

```cpp
if (isDroppable) {
    LOG_DEBUG("Dropping frame with flag: " << GST_BUFFER_FLAGS(buf) << ", PTS: " << buf->pts << ", DTS: " << buf->dts << " for " << kvssink->stream_name);
    goto CleanUp;
}
```

Sorry if my question is trivial.

**Reply by jdelapla** [reply]
Yes, this is what I was requesting. Please generate logs with this change and see if it's the PTS or DTS being invalid that is causing the dropped frames.

Edit: I saw your logs were already posted in the next comment, please review my suggestion.

**Comment by floatingpurr** [comment]
Ok, I have just run the command with the customized build with info about pts and dts for the camera foo.

```sh
gst-launch-1.0 -v rtspsrc protocols=tcp location="rtsp://foo" short-header=TRUE ! rtph264depay ! h264parse ! kvssink stream-name="YourStreamName" storage-size=128 access-key="foo-credential" secret-key="foo" aws-region="eu-central-1"
```

Here is the complete log: [foo_tcp_pts_dts.log](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/files/11429436/foo_tcp_pts_dts.log)


It looks like no frame is put into the Kinesis stream.

**Reply by jdelapla** [reply, marked as answer]
Okay, so I went through the logs and as suspected it is a timestamp corruption issue.

```2023-05-09 09:18:27 [140627329136192] DEBUG - Dropping frame with flag: 1536, PTS: 208075878, DTS: 18446744073709551615 for YourStreamName```

The DTS value is a max int value. If this is not correctable by the source then you can edit this line in the kvssink:
https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/63537aaa4398d4879c3f5a41d1149878cbf0dad1/src/gstreamer/gstkvssink.cpp#L1288
And remove the check on the DTS.

so this line should now be something like:
```
                        (GST_BUFFER_FLAG_IS_SET(buf, GST_BUFFER_FLAG_HEADER) && (!GST_BUFFER_PTS_IS_VALID(buf));
```

Let me know if this works.

**Reply by floatingpurr** [reply]
... aaand it works! Thank you very much for your kind help!

PS: I'm not an expert of RTSP or video streaming, what does the relaxing of the timestamp check imply?

PPS: any chance of having this option backed into the _kvssink_ options (e.g,. _check-dts_)?

**Reply by jdelapla** [reply]
Yeah I'm planning to add the option to disable this in the next release. The reason this behavior is in place by default is because decoding timestamps are used internally to determine the sequential ordering of frames in the event that frames were delivered out of order for any reason.

If a stream has been accepting valid DTS values and then received a MAX_INT value like the DTS in your stream, then it would end up assuming all subsequent frames were invalid.

However in your case ALL the DTS values are invalid. There is actually a nice substitution in the kvssink a few lines below this droppable check that allows for this use case to behave normally. However if for whatever reason your RTSP was to suddenly start giving DTS values that are valid, then your stream would stop working.

TL;DR, either every DTS needs to be valid, or every DTS needs to be invalid. Otherwise the sink won't work. The implication of removing this check is the assumption that all frames will have an invalid DTS, and all frames will be sent to kvssink in order.

**Reply by LanderN** [reply]
Just FYI @jdelapla , I ran into the same issue and your suggestion fixed it. I don't see an option to disable the DTS check in the current release, is that correct?

**Comment by marmarte-117** [comment]
Hi @jdelapla ! Are there any news about the possibility to enable or disable the DTS check? I'm also facing the same issue. Thanks!
